numpy/core/fromnumeric.py

```diff
@@ -1404,20 +1411,24 @@ def resize(a, new_shape):
     """
     if isinstance(new_shape, (int, nt.integer)):
         new_shape = (new_shape,)
+
     a = ravel(a)
-    Na = len(a)
-    total_size = um.multiply.reduce(new_shape)
-    if Na == 0 or total_size == 0:
-        return mu.zeros(new_shape, a.dtype)
 
-    n_copies = int(total_size / Na)
-    extra = total_size % Na
+    new_size = 1
+    for dim_length in new_shape:
+        new_size *= dim_length
+        if dim_length < 0:
+            raise ValueError('all elements of `new_shape` must be non-negative')
+
+    if a.size == 0 or new_size == 0:
+        # First case must zero fill. The second would have repeats == 0.
+        return np.zeros_like(a, shape=new_shape)
 
-    if extra != 0:
-        n_copies = n_copies + 1
-        extra = Na - extra
+    # How often do we have to repeat `a` and how much longer will it be?
+    extra = new_size % a.size
+    repeats = (new_size - extra) // a.size
```

Comment:
Shorter and perhaps faster as
```
repeats, extra - divmod(new_size, a.size)
```

Reply:
There is no down rounding here, since the result is always an exact integer (the `//` only makes sure it remains an int)? But yes, divmod is better, somehow forgot it is a python function/operator...

Reply:
There is rounding down, you did it via the subtraction.

As I read it, this will make `np.arange(5).resize(7)` give an array of the wrong size - your division will give `1`, and then extra will be `2` so you'll return an array of length `3`.

Reply:
Sorry, yeah, I guess I would have thought such a big error gets found by the tests :/

Reply:
Jeesh... emberassing. I just opted for duplicating the `concatenate` line to keep it stupid, open for other things, but...

Reply:
Is there a test for `np.arange(5).resize(7)`?

Reply:
There is now...

Reply:
where?